### PR TITLE
[BEAM-7079] Add Chicago Taxi Example running on Dataflow

### PR DESCRIPTION
**Pull request by mwalenia**
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run Python PreCommit

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Retest this please

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run Python PreCommit

**Comment by mwalenia** [comment]
R: @angoenka 
Ankur, could you take a look at this? Thanks!

**Comment by mwalenia** [comment]
Run python PreCommit

**Comment by mwalenia** [comment]
Run Python PreCommit

**Comment by mwalenia** [comment]
Run Python PreCommit

**Comment by mwalenia** [comment]
@angoenka Hi, could you take a look at this?
CC: @pabloem

**Comment by angoenka** [comment]
Thanks for the changes.
I suppose most of the pipeline code is from the chicago taxi example on TFX git repo. Instead of copying the code, can we some how dynamically pull the code from the TFX repo and run the pipeline. 
We can pin the pulled code to versions so that we are not testing against TFX master.

If not, we need to have a strategy to update these pipelines.

**Comment by mwalenia** [comment]
@angoenka thanks for the reply.

That's correct, most of the code comes from the TFX repo. However, I believe that it doesn't make sense to use unchanged code from there to run those tests - it would be impossible to gather metrics using Metrics API, which would make the results not representative of real Beam performance in this scenario. 

Also, the recommended way of running the example from the TFX repo is to use Airflow to orchestrate everything, which would create another big dependency that would need to be set up and torn down automatically on Jenkins.

I think that this approach is the most beneficial. I agree that if the pipelines in TFX example change or if we drop Python 2 support, it will be necessary to update the code, but I think this won't be very time-consuming.

WDYT?

**Comment by angoenka** [comment]
I see, so we do have changes to the pipeline for publishing metrics.
In that case, I agree that pulling the code from TFX repo will have to be modified or we have to comeup with some clever technique to publish metrics.
I think we can go ahead with the current state of the code.

**Comment by angoenka** [comment]
LGTM
Thanks!
Please make sure that the tests pass and then we can merge the changes.

**Comment by pabloem** [comment]
Run Python PreCommit

**Comment by mwalenia** [comment]
Run Python PreCommmit

**Comment by mwalenia** [comment]
Run Python PreCommmit

**Comment by mwalenia** [comment]
Run Portable_Python PreCommit

**Comment by mwalenia** [comment]
Run Python_PVR_Flink PreCommit

**Comment by mwalenia** [comment]
Run Portable_Python PreCommit

**Comment by mwalenia** [comment]
Run Python PreCommit

**Comment by mwalenia** [comment]
Run Python_PVR_Flink PreCommit

**Comment by pabloem** [comment]
Run Python PreCommit

**Comment by pabloem** [comment]
Run Python PreCommit

**Comment by mwalenia** [comment]
Run Python PreCommit

**Comment by mwalenia** [comment]
Run Python PreCommit

**Comment by mwalenia** [comment]
Run Python PreCommit

**Comment by mwalenia** [comment]
Run Python PreCommit

**Comment by pabloem** [comment]
Run Python PreCommit

**Comment by kamilwu** [comment]
Run Python PreCommit

**Comment by pabloem** [comment]
Run Python PreCommit

**Comment by pabloem** [comment]
Some odd issue with pip.

**Comment by mwalenia** [comment]
Run Python PreCommit

File: sdks/python/setup.py [hunk 1]
```diff
@@ -135,6 +135,15 @@ def get_version():
     'parameterized>=0.6.0,<0.7.0',
     'pyhamcrest>=1.9,<2.0',
     'tenacity>=5.0.2,<6.0',
+    'jupyter>=1.0,<2',
```
[the remark this answers is not in the document]
Comment (mwalenia) [reply sdks/python/setup.py]: I will look into it, thanks!

File: sdks/python/build.gradle [hunk 2]
```diff
@@ -168,6 +168,21 @@ task portableWordCount {
...
+  
+  def gcsRoot = findProperty('gcsRoot')
+  def runner = findProperty('runner')
+  def cliArgs = "${gcsRoot} ${runner} ${project.buildDir}/apache-beam.tar.gz"
+  
+  doLast {
+    exec {
+      workingDir "$project.projectDir/apache_beam/testing/benchmarks/chicago_taxi/"
+      executable 'sh'
+      args '-c', ". ${project.ext.envdir}/bin/activate && ./run_chicago.sh ${cliArgs}"
```
[the remark this answers is not in the document]
Comment (mwalenia) [reply sdks/python/build.gradle]: Thank you for the suggestion, I applied it. WDYT?

**Comment by mwalenia** [comment]
I rebased the branch onto master in order to get Python PreCommit working again

**Comment by pabloem** [comment]
The failures are now docs and lint failures that seem related to this change.

**Comment by mwalenia** [comment]
@robertwb There's a problem with extracting the dependencies from the test dependency list in setup.py: linters need the dependencies in order to properly parse the files and moving the installation to a later stage breaks them - see [this PreCommit job](https://builds.apache.org/job/beam_PreCommit_Python_Commit/7723/consoleFull#gradle-task-148), especially the docs task.

**Comment by robertwb** [comment]
The lint errors look unrelated (due to placement of __future__ imports wrt
other code).

As for docs, you can exclude chicago taxi in the list at
scripts/generate_pydoc.sh. This is what we should do regardless, as it
doesn't make sense to publish these as part of our python reference
documentation.

**Comment by mwalenia** [comment]
Run Python_PVR_Flink PreCommit

**Comment by mwalenia** [comment]
@robertwb Thanks for the suggestions, disabling doc task helped.

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Chicago Taxi on Dataflow

**Comment by mwalenia** [comment]
@angoenka Everything passes now and I implemented Robert's suggestions. It looks ready to merge, WDYT?

**Comment by pabloem** [comment]
Thanks all : )

**Pull request merged**